### DISCUSSION TITLE
Can inline assembly in sycl like cuda does?

### DISCUSSION THREAD

**Discussion by alanzhai219** [intel/llvm#13203] [Q&A] [answered]
You know, developers can inline PTX in the cuda kernel, which is much more efficient for expert developers. That is much more important for extreme optimization.
Can DPC++ or sycl support inline assembly?
If yes, please provide a sample.
If not, is any plan for such feature.
Thank you.

**Comment by JackAKirk** [comment]
Yes it is exactly as in the nvcc compiler (see https://docs.nvidia.com/cuda/inline-ptx-assembly/index.html) except you have to wrap ptx code in the following MACRO:

```
if defined(__NVPTX__)
your_ptx_asm_here
#endif
```

In a parallel_for.

**Reply by alanzhai219** [reply]
@JackAKirk  Maybe my description is unclear. I mean SYCL can support inline assembly now?

**Reply by AlexeySachkov** [reply, marked as answer]
@alanzhai219, yes, this functionality is supported already, you can find some examples (but with Intel GPUs ASM) here: https://github.com/intel/llvm/tree/sycl/sycl/test-e2e/InlineAsm

**Reply by JackAKirk** [reply]
Yes, I mean to do 
```
if defined(__NVPTX__)
your_ptx_asm_here
#endif
```

inside any SYCL parallel_for. e.g. `queue::parallel_for`.

**Reply by alanzhai219** [reply]
@AlexeySachkov Thanks, that is what I want. BTW, is there any such asm spec for intel GPU?

**Reply by AlexeySachkov** [reply]
@alanzhai219, the only document I'm aware of is https://github.com/intel/intel-graphics-compiler/blob/master/documentation/visa/1_introduction.md

**Reply by alanzhai219** [reply]
@AlexeySachkov  Thanks a lot.
